components/Heading/index.tsx

```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+
+export type Depth = 1 | 2 | 3 | 4 | 5 | 6;
+
+const CreateHeading = (depth: Depth) => {
+  const Tag: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' = `h${depth}`;
+
+  const Heading = ({ id, children, ...attrs }: React.PropsWithChildren<HTMLHeadingElement>) => {
+    if (!children) return '';
+
+    return (
+      // @ts-expect-error
+      <Tag {...attrs} className={`heading heading-${depth} header-scroll`}>
+        <div key={`heading-anchor-${id}`} className="heading-anchor anchor waypoint" id={id} />
+        <div key={`heading-text-${id}`} className="heading-text">
+          {children}
+        </div>
+        <a
+          key={`heading-anchor-icon-${id}`}
+          aria-label={`Skip link to ${children[1]}`}
+          className="heading-anchor-icon fa fa-anchor"
+          href={`#${id}`}
+        />
+      </Tag>
+    );
+  };
+
+  return Heading;
+};
```

Comment:
I think we should always prefer to keep component definitions at the top-level where we can, rather than nested them within a function:

```suggestion
const Heading = ({ tag: Tag = 'h3', depth = 3, id, children, ...attrs }: React.PropsWithChildren<HTMLHeadingElement>) => {
  if (!children) return '';

  return (
    <Tag {...attrs} className={`heading heading-${depth} header-scroll`}>
      <div key={`heading-anchor-${id}`} className="heading-anchor anchor waypoint" id={id} />
      <div key={`heading-text-${id}`} className="heading-text">
        {children}
      </div>
      <a
        key={`heading-anchor-icon-${id}`}
        aria-label={`Skip link to ${children[1]}`}
        className="heading-anchor-icon fa fa-anchor"
        href={`#${id}`}
      />
    </Tag>
  );
};

const CreateHeading = (depth: Depth) => props => <Heading {...props} depth={depth} Tag={`h${depth}`} />;
```